main.py

```diff
@@ -1,38 +1,52 @@
+import argparse
 import multiprocessing
 import ssl
+import os
 from typing import List, Callable
+from str2bool import str2bool
 
-import downloader
-import packager
-import vscode
 
-def __get_lines(file_path: str) -> List[str]:
-    with open(file_path, 'r') as file:
-        return list(filter(lambda x: len(x) > 0, file.read().splitlines()))
+from src.providers.Vscode import Vscode
+from src.providers.Npm import Npm
+from src.products_reader import get_lines
+from src.downloader import download
+from src.packager import package
 
 
-def __download_single_package(base_directory: str, registry_name: str, package_name: str, package_ext: str, package_getter: Callable) -> str:
-    (version_name, url) = package_getter(package_name)
-    return downloader.download(base_directory, registry_name, package_name, version_name, package_ext, url)
+parser = argparse.ArgumentParser(description='CLI tool')
+parser.add_argument('--vscode', type=str2bool, default=False,
+                    help='specify if should download vscode extentions')
+parser.add_argument('--npm', type=str2bool, default=True,
+                    help='specify if should download npm packages')
+parser.add_argument('--concurrency', type=int, default=2,
+                    help='Number of workers = concurrency * CPU_CORES_COUNT')
+parser.add_argument('--proxy', type=str2bool, default=False,
+                    help='Turn on if you use a proxy')
+args = parser.parse_args()
 
 
-def __download_packages(base_directory: str, registry_name: str, package_ext: str, package_getter: Callable, package_list_file: str) -> List[str]:
-    package_names = [
-        (base_directory, registry_name, package_name, package_ext, package_getter)
-        for package_name in  __get_lines(package_list_file)
-    ]
-    with multiprocessing.Pool(16) as pool:
-        downloaded = pool.starmap(__download_single_package, package_names)
-    return downloaded
-
- 
-ssl._create_default_https_context = ssl._create_unverified_context
-
 def main() -> None:
     base_directory = './packages'
-    downloaded = __download_packages(base_directory, 'vscode', 'vsix', vscode.get_extension, 'vscode.list')
-    # ToDo save to some (file based?) DB all downloaded files and clean downloads
-    packager.package(downloaded)
+    resolved_products = []
+    num_workers = args.concurrency * os.cpu_count()
+    if args.proxy:
+        ssl._create_default_https_context = ssl._create_unverified_context
+    if args.npm:
+        npm = Npm()
```

Comment:
this is code repetition. isn't the interface meant to avoid this?

Reply:
you mean the `if`s are code repetition?

Reply:
Yes

Reply:
the whole point of having the interface is sending all the providers to work at once, I agree that if 3 more minutes we can avoid these `if`s, I will do that

Reply:
please revisit the code